tsconsensus/tsconsensus_test.go

```diff
@@ -0,0 +1,697 @@
+func assertCommandsWorkOnAnyNode(t *testing.T, participants []*participant) {
+	for i, p := range participants {
+		res, err := p.c.ExecuteCommand(Command{Args: []byte{byte(i)}})
+		if err != nil {
+			t.Fatalf("%d: Error ExecuteCommand: %v", i, err)
+		}
+		if res.Err != nil {
+			t.Fatalf("%d: Result Error ExecuteCommand: %v", i, res.Err)
+		}
+		retVal := int(res.Result[0])
+		// the test implementation of the fsm returns the count of events that have been received
+		if retVal != i+1 {
+			t.Fatalf("Result, want %d, got %d", i+1, retVal)
+		}
+
+		expectedEventsLength := i + 1
+		fxEventsInAll := func() bool {
+			for _, pOther := range participants {
+				if pOther.sm.numEvents() != expectedEventsLength {
+					return false
+				}
+			}
+			return true
+		}
+		waitFor(t, "event makes it to all", fxEventsInAll, 10, time.Second*1)
+	}
+}
+
+func TestConfig(t *testing.T) {
+	nettest.SkipIfNoNetwork(t)
+	ctx, cancel := context.WithTimeout(context.Background(), testContextTimeout)
+	defer cancel()
+	clusterTag := "tag:whatever"
+	ps, _, _ := startNodesAndWaitForPeerStatus(t, ctx, clusterTag, 3)
+	cfg := warnLogConfig()
+	// test all is well with non default ports
+	cfg.CommandPort = 12347
+	cfg.RaftPort = 11882
+	mp := uint16(8798)
+	cfg.MonitorPort = mp
+	createConsensusCluster(t, ctx, clusterTag, ps, cfg, true)
+	for _, p := range ps {
+		defer p.c.Stop(ctx)
+	}
+	assertCommandsWorkOnAnyNode(t, ps)
+
+	url := fmt.Sprintf("http://%s:%d/", ps[0].c.self.host, mp)
+	httpClientOnTailnet := ps[1].ts.HTTPClient()
+	rsp, err := httpClientOnTailnet.Get(url)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if rsp.StatusCode != 200 {
+		t.Fatalf("monitor status want %d, got %d", 200, rsp.StatusCode)
+	}
+	defer rsp.Body.Close()
+	body, err := io.ReadAll(rsp.Body)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Not a great assertion because it relies on the format of the response.
+	line1 := strings.Split(string(body), "\n")[0]
+	if line1[:10] != "RaftState:" {
+		t.Fatalf("getting monitor status, first line, want something that starts with 'RaftState:', got '%s'", line1)
+	}
+}
+
+func TestFollowerFailover(t *testing.T) {
+	nettest.SkipIfNoNetwork(t)
+	ctx, cancel := context.WithTimeout(context.Background(), testContextTimeout)
+	defer cancel()
+	clusterTag := "tag:whatever"
+	ps, _, _ := startNodesAndWaitForPeerStatus(t, ctx, clusterTag, 3)
+	cfg := warnLogConfig()
+	createConsensusCluster(t, ctx, clusterTag, ps, cfg, false)
+	for _, p := range ps {
+		defer p.c.Stop(ctx)
+	}
+
+	smThree := ps[2].sm
+
+	fut := ps[0].c.raft.Apply([]byte("a"), 2*time.Second)
+	futTwo := ps[0].c.raft.Apply([]byte("b"), 2*time.Second)
+	err := fut.Error()
+	if err != nil {
+		t.Fatalf("Apply Raft error %v", err)
+	}
+	err = futTwo.Error()
+	if err != nil {
+		t.Fatalf("Apply Raft error %v", err)
+	}
+
+	fxAllMachinesHaveTheApplies := func() bool {
+		return ps[0].sm.numEvents() == 2 && ps[1].sm.numEvents() == 2 && smThree.numEvents() == 2
+	}
```

Comment:
Is checking the count sufficient or should you verify the contents and that they're in the correct order?